GH-14932: [Python] Add python bindings for JSON streaming reader

pan-x-c opened the pull request:
### Rationale for this change

The C++ arrow has a JSON streaming reader which is not exposed on the Python interface.

### What changes are included in this PR?

This PR is based on #33761. It adds the `open_json` method to open a streaming reader for a JSON file.

### Are these changes tested?

Yes

### Are there any user-facing changes?

Yes. A new `open_json` method has been added to the Python interface, located at `pyarrow.json.open_json`, and its parameters are the same as the `pyarrow.json.read_json`

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->
* GitHub Issue: #14932

pan-x-c commented on python/pyarrow/tests/test_json.py:
```diff
@@ -339,6 +334,300 @@ def test_stress_block_sizes(self):
+        assert reader.schema == expected_schema
+        assert reader.read_next_batch().to_pydict() == expected_data
+
+    def test_reconcile_across_blocks(self):
```
This original test case in `BaseTestJSON.test_reconcile_across_blocks` is not compatible with the JSON stream reader. 
I made some changes, but I don't know whether this change is reasonable.

pitrou commented on python/pyarrow/_json.pyx:
```diff
@@ -265,6 +267,37 @@ cdef _get_parse_options(ParseOptions parse_options, CJSONParseOptions* out):
+        Schema schema
+
+    def __init__(self):
+        raise TypeError("Do not call {}'s constructor directly, "
+                        "use pyarrow.json.open_json() instead."
+                        .format(self.__class__.__name__))
```
Nit: we can use f-strings now
```suggestion
        raise TypeError(f"Do not call {self.__class__.__name__}'s "
                        "constructor directly, "
                        "use pyarrow.json.open_json() instead.")
```

pitrou commented on python/pyarrow/_json.pyx:
```diff
@@ -265,6 +267,37 @@ cdef _get_parse_options(ParseOptions parse_options, CJSONParseOptions* out):
     else:
         out[0] = parse_options.options
 
+cdef class JSONStreamingReader(RecordBatchReader):
```
Should probably have another line break here to satisfy PEP 8 requirements.
```suggestion

cdef class JSONStreamingReader(RecordBatchReader):
```

pitrou commented on python/pyarrow/_json.pyx:
```diff
@@ -308,3 +341,45 @@ def read_json(input_file, read_options=None, parse_options=None,
+        Options for the JSON parser
+        (see pyarrow.json.ParseOptions constructor for defaults)
+    memory_pool : MemoryPool, optional
+        Pool to allocate Table memory from
```
```suggestion
        Pool to allocate RecordBatch memory from
```

pitrou commented on python/pyarrow/tests/test_json.py:
```diff
@@ -111,26 +112,20 @@ def test_parse_options(pickle_module):
+        """
+        :param b: bytes to be parsed
+        :param kwargs: arguments passed on to open the json file
+        :return: b parsed as a single RecordBatch
```
```suggestion
        :return: b parsed as a single Table
```

pitrou commented on python/pyarrow/tests/test_json.py:
```diff
@@ -357,3 +608,17 @@ def read_json(self, *args, **kwargs):
+
+
+class TestSerialStreamingJSONRead(BaseTestStreamingJSONRead, unittest.TestCase):
+
+    @property
+    def use_threads(self):
+        return False
```
Nit: I think this can simply be:
```suggestion
    use_threads = False
```

pitrou commented on python/pyarrow/tests/test_json.py:
```diff
@@ -357,3 +608,17 @@ def read_json(self, *args, **kwargs):
+
+@pytest.mark.threading
+class TestThreadedStreamingJSONRead(BaseTestStreamingJSONRead, unittest.TestCase):
+    @property
+    def use_threads(self):
+        return True
```
Same here

pitrou reviewed (changes requested):
This looks mostly excellent @pan-x-c , thank you.

Here are a couple minor comments.

Also, could you add the new API to the docs in https://github.com/apache/arrow/blob/main/docs/source/python/api/formats.rst#json-files and perhaps mention it in https://github.com/apache/arrow/blob/main/docs/source/python/json.rst ?

pan-x-c commented:
I have fixed the above comments and some similar issues in the `csv` module. I also added a brief introduction to `open_json` based on the doc of `open_csv`.

pitrou reviewed (changes requested):
Thanks a lot for the update, one comment remaining below.

pan-x-c commented on docs/source/python/json.rst:
```diff
@@ -115,3 +115,20 @@ and pass it to :func:`read_json`.  For example, you can pass an explicit
+There are a few caveats:
+
+1. For now, the incremental reader is always single-threaded (regardless of
+   :attr:`ReadOptions.use_threads`)
```
Done, but I'm not sure if the doc for `open_csv` has the same problem.

pitrou replied:
From a quick look, the CSV streaming reader is not multi-threaded.

pitrou reviewed (approved):
Thanks a lot for your contribution @pan-x-c ! This looks good to me now, I will merge if CI is green.
